### ISSUE TITLE
I want Podam to return the current time when producing java.time.LocalDateTime

### ISSUE THREAD

**Issue by telmperez** [mtedone/podam#104]
Just tried getting Podam to generate LocalDateTime.java objects (new Java 8) and I'm having some problems. Tried using an ExternalFactory without any luck (seeing strange runtime issues). Only thing that works right now is @PodamStrategyValue.

**Comment by mtedone** [comment]
Please attach the POJO you're trying to fill

**Comment by telmperez** [comment]
Absolutely!

Here's a class with a test, 

the output is;
Correct-> 2015-05-18T01:54:52.809
2015-05-18T08:35:10.376933995
2015-05-18T10:06:05.472141998
2015-05-18T04:43:58.582282988
+7181996-09-01T12:56:42.172240635
+7182022-09-06T10:05:39.060489072
2015-05-18T02:32:26.744446803
2015-05-18T15:08:46.836505757
+7182063-09-06T06:07:54.812981006
2015-05-18T14:39:40.922392669
2015-05-18T15:57:16.055973486

Thanks!

```
import org.junit.Test;
import uk.co.jemos.podam.api.PodamFactory;
import uk.co.jemos.podam.api.PodamFactoryImpl;

import java.time.LocalDateTime;

public class PodamDateTimeTest {

  public class TestPojo {
    private LocalDateTime localDateTime;

    public LocalDateTime getLocalDateTime() {
      return localDateTime;
    }

    public void setLocalDateTime(LocalDateTime localDateTime) {
      this.localDateTime = localDateTime;
    }
  }

  @Test
  public void dateTimeTest() {
    PodamFactory podamFactory = new PodamFactoryImpl();
    podamFactory.getStrategy().setMemoization(false);

    System.out.println("Correct-> " + LocalDateTime.now());

    for (int i = 0; i < 10; i++) {
      TestPojo testPojo = podamFactory.manufacturePojo(TestPojo.class);
      System.out.println(testPojo.getLocalDateTime());
    }
  }

}
```

**Comment by telmperez** [comment]
When I added an external Podam Factory I noticed that the manufacturePojo function was being called a different number of times when I used an if condition,

```pojoClass.isAssignableFrom(LocalDateTime.class)```

There were several different interfaces, including the Clock interface that was sometimes populated,

```
@Override
  public <T> T manufacturePojo(Class<T> pojoClass, Type... genericTypeArgs) {
    LOG.info(pojoClass.getCanonicalName());
    try {
      if (pojoClass.isAssignableFrom(LocalDateTime.class)) {
        LOG.info(pojoClass.getCanonicalName());
        @SuppressWarnings("unchecked")
        T localDateTime = (T) LocalDateTime.now();
        return localDateTime;
      }
    }
    catch (Exception e) {
      throw new IllegalStateException("Manufacturing failed", e);
    }
    return null;
  }
```

**Comment by mtedone** [comment]
Since PodamStrategyValue works this is not a bug. We'll be looking at the
external factory but we can't introduce Java 8 native classes in Podam as
it needs to be backwards compatible. There are users who are using it and
are still in Java 6. At some point in the next few months we'll create a
version which requires Java 8 and at that point we'll be able to introduce
native support for Java 8 date and time classes.

Regards,

**Comment by daivanov** [comment]
I did change title of issue to reflect real issue as I was able it to understand from this thread. The issue is invalid, because by default Podam is using RandomDataProviderStrategy, which produces random objects, so expecting the current time stamp is wrong. You should replace RandomDataProviderStrategy with your own implementation, which will be filling LocalDateTime with local date. External factory doesn't seem to be valid approach here, as it gets called only, if object could not be produced by Podam, but LocalDateTime can be produced by Podam, so external factory won't be called.

**Comment by mtedone** [comment]
I'll create an "Invalid" label and then we can close this issue.

**Comment by telmperez** [comment]
Thanks for the quick response. Agreed that the title was incorrect, I updated the title again to reflect this response, please feel free to update it again.

I think the issue here is not that it isn't returning the current date time it is returning a date that could potentially break downstream frameworks considering that the date's being generated in my test are in the year 23 Million or the current date, which is why I originally filed the issue.

I did replace the RandomDataProviderStrategy, however I was unable to find a way to add a default Class or strategy for known types. The only thing I could find was the ability to use the "addSpecific" function. I wasn't able to find a way to add a default AttributeStrategy for a particular class type (which would be very useful). Please let me know if there's another way to do this using the RandomDataProviderStrategy.

Also I can definitely confirm that the External factory is called when trying to manufacture a LocalDateTime object.

Thanks again! Been using Podam for a few years now without issues, you guys rock!

**Comment by telmperez** [comment]
Updated the title again since the default Podam behaviour appears to be creating the current date when generating java.util.Date objects. This behaviour should be the same when dealing with the new Java 8 dates.

**Comment by daivanov** [comment]
> returning a date that could potentially break downstream frameworks considering that the date's being generated in my test are in the year 23 Million or the current date, which is why I originally filed the issue.

This date is an absolutely valid date. If it breaks framework, then Podam did a good job, found a bug and bug must be fixed in framework.

> I did replace the RandomDataProviderStrategy, however I was unable to find a way to add a default Class or strategy for known types.

You probably want to reverse sorting order in method comparator just like I did it.

**Comment by mtedone** [comment]
OK, we'll provide more consistency around the provision of date objects but this is not a bug, since the two objects are different and therefore have different internals.

**Comment by telmperez** [comment]
Thanks! I'm not 100% certain if it's a bug or not. My concerns are that LocalDateTime can't really be created using reflection since it's made up of several interfaces which could conflict with one another depending on how they are stitched together. I noticed different classes (interfaces) being created (internally within Podam) when I looked at what was happening when a LocalDateTime object was manufactured.

I did write a more detailed test that generated millions of date objects finding some interesting usages for dates but I could not find the problem. I did however find one bug with hsqldb not being able to handle dates with really large years.

**Comment by telmperez** [comment]
Regarding your previous comment, I had a look at the sort() functions on the DataProviderStrategy interface and I couldn't figure out how the sort function could help me in this situation. Do you have that example available? Thanks again!

**Comment by daivanov** [comment]
> LocalDateTime can't really be created using reflection since it's made up of several interfaces

What does it mean? It contradicts to your second comment and result of my tests.

> I did however find one "issue" with hsqldb not being able to handle dates with podam generated objects.

Report this to HSQLDB.

**Comment by daivanov** [comment]
Fixed in 5.4.1-SNAPSHOT
